File: app/src/main/java/com/example/trigger/ImagePreference.java
package com.example.trigger;

import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Icon;
import android.preference.Preference;
import android.preference.SwitchPreference;
import android.util.AttributeSet;

import com.example.trigger.SetupActivity;
import com.jcraft.jsch.KeyPair;

import android.graphics.Bitmap;

/*
* Stores a string in preference. Indicated the presence of data with a switch.
* On click, the file chooser opens.
*/
public class ImagePreference extends SwitchPreference {
    private Bitmap image;
    private Context context;
    static ImagePreference self;

    public ImagePreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        this.context = context;
        final ImagePreference self = this;

        // tell the superclass that we handle the value on out own!
        setPersistent(false);

        if (this.image == null) {
            setChecked(false);
        } else {
            setChecked(true);
        }

        this.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                if (self.image == null) {
                    self.setChecked(false);
                } else {
                    self.setChecked(true);
                }
                
                return false;
            }
        });

        this.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                //String register_url = ((SetupActivity) self.context).getRegisterUrl();

                // store in public static member - hack!
                ImagePreference.self = ImagePreference.this;

                Intent intent = new Intent(self.context, ImageActivity.class);
                //intent.putExtra("register_url", register_url);
                self.context.startActivity(intent);

                return false;
            }
        });
    }

    public void setImage(Bitmap image) {
        this.image = image;

        if (this.image == null) {
            setChecked(false);
        } else {
            setChecked(true);
        }
    }

    public Bitmap getImage() {
        return image;
    }

    @Override
    protected void onSetInitialValue(boolean restorePersistedValue, Object defaultValue) {
        if (image == null) {
            super.onSetInitialValue(restorePersistedValue, false);
        } else {
            super.onSetInitialValue(restorePersistedValue, true);
        }
    }
}

File: app/src/main/kotlin/app/trigger/ImagePreference.kt
package app.trigger

import android.graphics.Bitmap
import android.content.*
import android.preference.SwitchPreference
import android.preference.Preference.OnPreferenceChangeListener
import android.preference.Preference.OnPreferenceClickListener
import android.util.AttributeSet

/*
* Stores a string in preference. Indicated the presence of data with a switch.
* On click, the file chooser opens.
*/
class ImagePreference(private val context_: Context, attrs: AttributeSet?) : SwitchPreference(context_, attrs) {
    var image: Bitmap? = null
        set(value) {
            field = value
            isChecked = (value != null)
        }

    override fun onSetInitialValue(restorePersistedValue: Boolean, defaultValue: Any) {
        super.onSetInitialValue(restorePersistedValue, (image != null))
    }

    companion object {
        var self: ImagePreference? = null
    }

    init {
        val self = this

        // tell the superclass that we handle the value on out own!
        isPersistent = false
        isChecked = (image != null)
        this.onPreferenceChangeListener = OnPreferenceChangeListener { preference, newValue ->
            self.isChecked = (self.image != null)
            false
        }
        this.onPreferenceClickListener = OnPreferenceClickListener { //String register_url = ((SetupActivity) self.context).getRegisterUrl();

            // store in public static member - hack!
            Companion.self = this@ImagePreference
            val intent = Intent(self.context, ImageActivity::class.java)
            //intent.putExtra("register_url", register_url);
            self.context.startActivity(intent)
            false
        }
    }
}
